libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/transforms/ClasspathEntrySnapshotTransform.kt

```diff
@@ -7,25 +7,65 @@ package org.jetbrains.kotlin.gradle.internal.transforms
 
 /** Transform to create a snapshot of a classpath entry (directory or jar). */
 @CacheableTransform
-abstract class ClasspathEntrySnapshotTransform : TransformAction<TransformParameters.None> {
+abstract class ClasspathEntrySnapshotTransform : TransformAction<ClasspathEntrySnapshotTransform.Parameters> {
+
+    abstract class Parameters : TransformParameters {
+        @get:Internal
+        abstract val buildMetricsReporterService: Property<BuildMetricsReporterService>
+    }
 
     @get:Classpath
     @get:InputArtifact
     abstract val inputArtifact: Provider<FileSystemLocation>
 
     override fun transform(outputs: TransformOutputs) {
-        val classpathEntry = inputArtifact.get().asFile
-        val snapshotFile = outputs.file(classpathEntry.name.replace('.', '_') + "-snapshot.bin")
+        val classpathEntryInputDirOrJar = inputArtifact.get().asFile
+        val snapshotOutputFile = outputs.file(classpathEntryInputDirOrJar.name.replace('.', '_') + "-snapshot.bin")
+        val buildMetricsReporterService = parameters.buildMetricsReporterService.orNull
+
+        val metricsReporter = buildMetricsReporterService?.let { BuildMetricsReporterImpl() } ?: DoNothingBuildMetricsReporter
+        val startTimeMs = System.currentTimeMillis()
+        var failureMessage: String? = null
+        try {
+            doTransform(classpathEntryInputDirOrJar, snapshotOutputFile, metricsReporter)
+        } catch (e: Throwable) {
+            failureMessage = e.message
+            throw e
+        } finally {
+            buildMetricsReporterService?.addTransformMetrics(
+                transformPath = "${ClasspathEntrySnapshotTransform::class.simpleName} for ${classpathEntryInputDirOrJar.path}",
+                transformClass = ClasspathEntrySnapshotTransform::class.java,
+                isKotlinTransform = true,
+                startTimeMs = startTimeMs,
+                totalTimeMs = System.currentTimeMillis() - startTimeMs,
+                buildMetrics = metricsReporter.getMetrics(),
+                failureMessage = failureMessage
+            )
+        }
+    }
 
-        val snapshot = ClasspathEntrySnapshotter.snapshot(classpathEntry)
-        ClasspathEntrySnapshotExternalizer.saveToFile(snapshotFile, snapshot)
+    private fun doTransform(classpathEntryInputDirOrJar: File, snapshotOutputFile: File, metrics: BuildMetricsReporter) {
+        metrics.measure(BuildTime.CLASSPATH_ENTRY_SNAPSHOT_TRANSFORM) {
+            val snapshot = ClasspathEntrySnapshotter.snapshot(classpathEntryInputDirOrJar, metrics)
+            metrics.measure(BuildTime.SAVE_CLASSPATH_ENTRY_SNAPSHOT) {
+                ClasspathEntrySnapshotExternalizer.saveToFile(snapshotOutputFile, snapshot)
+            }
+        }
+        metrics.addMetric(BuildPerformanceMetric.CLASSPATH_ENTRY_SNAPSHOT_TRANSFORM_EXECUTION_COUNT, 1)
+        metrics.addMetric(BuildPerformanceMetric.CLASSPATH_ENTRY_SIZE, classpathEntryInputDirOrJar.walk().sumOf { it.length() })
```

Comment:
This may add non-trivial overhead to transforms, if we need to walk dir tree. How do you plan to use this data, can we collect it only if explicitly requested?

Reply:
We can use this data to understand the project better, to detect if there is anything unusual or to guide some of our decisions (e.g., cache size).

I've moved these lines to inside `metrics.measure(BuildTime.CLASSPATH_ENTRY_SNAPSHOT_TRANSFORM) {` so any overhead will not go undetected.

We can probably remove this later if the overhead is too high or if we see that there's nothing to learn from the data collected.

Reply:
Can you please avoid adding overhead unless performance data is explicitly requested? Or collect data as part side-effect of class file processing?

We know for sure that this increases the transform time, so we try to avoid it.

Reply:
Okay I now collect the data only if it's a jar (which is by far the common case).